### PR TITLE
daemon, overlord/snapstate: give RevisionOptions a CohortKey

### PR DESCRIPTION
**Pull request by chipaca**
This means refresh and switch both gain cohort abilities.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1071,12 +1073,20 @@ func snapSwitch(inst *snapInstruction, st *state.State) (string, []*state.TaskSe
...
 
-	msg := fmt.Sprintf(i18n.G("Switch %q snap to %s"), inst.Snaps[0], inst.Channel)
+	var msg string
+	switch {
+	case inst.CohortKey == "" && inst.Channel != "":
+		msg = fmt.Sprintf(i18n.G("Switch %q snap to %s"), inst.Snaps[0], inst.Channel)
+	case inst.CohortKey != "" && inst.Channel == "":
+		msg = fmt.Sprintf(i18n.G("Switch %q snap to cohort %q"), inst.Snaps[0], strutil.ElliptRight(inst.CohortKey, 10))
+	default:
+		msg = fmt.Sprintf(i18n.G("Switch %q snap to channel %q and cohort %q"), inst.Snaps[0], inst.Channel, strutil.ElliptRight(inst.CohortKey, 10))
```
Comment (mvo5) [inline daemon/api.go]: Is this path testsed?

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1349,16 +1361,36 @@ func UpdateUnderDeviceContext(st *state.State, name string, opts *RevisionOption
...
-			// update the tracked channel
+			// update the tracked channel and cohort
 			snapsup.Channel = opts.Channel
+			snapsup.CohortKey = opts.CohortKey
 			// Update the current snap channel as well. This ensures that
 			// the UI displays the right values.
 			snapsup.SideInfo.Channel = opts.Channel
 
-			switchSnap := st.NewTask("switch-snap-channel", fmt.Sprintf(i18n.G("Switch snap %q from %s to %s"), snapsup.InstanceName(), snapst.Channel, opts.Channel))
+			switchSnap := st.NewTask("switch-snap-channel", summary)
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go]: Do we test the summary? It seems to make sense now that it gets constructed in a more "complicated" way.

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1337,7 +1349,7 @@ func UpdateUnderDeviceContext(st *state.State, name string, opts *RevisionOption
 	}
 
 	// see if we need to update the channel or toggle ignore-validation
-	if infoErr == store.ErrNoUpdateAvailable && (snapst.Channel != opts.Channel || snapst.IgnoreValidation != flags.IgnoreValidation) {
+	if infoErr == store.ErrNoUpdateAvailable && (snapst.Channel != opts.Channel || snapst.IgnoreValidation != flags.IgnoreValidation || snapst.CohortKey != opts.CohortKey) {
```
Comment (bboozzoo) [inline overlord/snapstate/snapstate.go]: maybe we could add helpers:
```
changeChannel := snapst.Channel != opts.Channel
changeCohortKey := snapst.CohortKey != opts.CohortKey
changeIgnoreValidation := snapst.IgnoreValidation != opts.IgnoreValidation
if infoErr == storeErrNoUpdateAvailable && (changeChannel || changeCohortKey || changeIgnoreValidation) {
....
```

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -1349,16 +1361,36 @@ func UpdateUnderDeviceContext(st *state.State, name string, opts *RevisionOption
 			SideInfo:    snapst.CurrentSideInfo(),
 			Flags:       snapst.Flags.ForSnapSetup(),
 			InstanceKey: snapst.InstanceKey,
-		}
+			CohortKey:   opts.CohortKey,
+		}
+
+		if snapst.Channel != opts.Channel || snapst.CohortKey != opts.CohortKey {
```
Comment (bboozzoo) [inline overlord/snapstate/snapstate.go]: Then here:
```
if changeChannel || chaneCohortKey {
    switch {
    case !changeChannel:
    ...
    case !changeCohortKey:
    ...
```

File: overlord/snapstate/snapmgr.go [hunk 5]
```diff
@@ -382,7 +382,7 @@ func Manager(st *state.State, runner *state.TaskRunner) (*SnapManager, error) {
 	runner.AddCleanup("copy-snap-data", m.cleanupCopySnapData)
 	runner.AddHandler("link-snap", m.doLinkSnap, m.undoLinkSnap)
 	runner.AddHandler("start-snap-services", m.startSnapServices, m.stopSnapServices)
-	runner.AddHandler("switch-snap-channel", m.doSwitchSnapChannel, nil)
+	runner.AddHandler("switch-snap-channel", m.doSwitchSnapChannel, nil) // also switches the cohort
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/snapmgr.go]: `doSwitchChannelAndOrCohort`?

File: overlord/snapstate/handlers.go [hunk 6]
```diff
@@ -1326,6 +1327,7 @@ func (m *SnapManager) doSwitchSnapChannel(t *state.Task, _ *tomb.Tomb) error {
 
 	// switched the tracked channel
 	snapst.Channel = snapsup.Channel
+	snapst.CohortKey = snapsup.CohortKey
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: shouldn't this be done only if snapsup.CohortKey != "" ?

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -1243,6 +1248,7 @@ func Switch(st *state.State, name string, opts *RevisionOptions) (*state.TaskSet
 	snapsup := &SnapSetup{
 		SideInfo:    snapst.CurrentSideInfo(),
 		Channel:     opts.Channel,
+		CohortKey:   opts.CohortKey,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: or this should be snapst.CohortKey if opts.CohortKey is ""

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -969,7 +973,8 @@ func doUpdate(ctx context.Context, st *state.State, names []string, updates []*s
 		snapsup := &SnapSetup{
 			Base:         update.Base,
 			Prereq:       defaultContentPlugProviders(st, update),
-			Channel:      updateOpts.Channel,
+			Channel:      revnoOpts.Channel,
+			CohortKey:    revnoOpts.CohortKey,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: this should be snapst.CohortKey if that revno.CohortKey is "" or the logic ins doLinkSnap etc needs to change

**Comment by pedronis** [comment]
to be clearer it seems we are missing a test that:

snap install --cohort=coo foo
snap refresh coo
snap refresh coo

will use the cohort for the two refreshes and keep it set

File: overlord/snapstate/snapmgr.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chipaca) [reply overlord/snapstate/snapmgr.go]: `doSwitchSnap`

File: overlord/snapstate/snapstate.go (hunk 7 above)
Comment (chipaca) [reply overlord/snapstate/snapstate.go]: yes.

File: overlord/snapstate/snapstate.go (hunk 8 above)
Comment (chipaca) [reply overlord/snapstate/snapstate.go]: it's defaulted to snapst.CohortKey in the params() func

Comment (chipaca) [reply overlord/snapstate/snapstate.go]: sorry, not in `parms()`; search for `// default to being in the same cohort`

Comment (chipaca) [reply overlord/snapstate/snapstate.go]: sigh. Final clarification: for single snap update, it's defaulted to snapst.CohortKey if opts.CohortKey is unset. In multi-snap update, opts.CohortKey is ignored and it's set to snapst.CohortKey.

File: overlord/snapstate/snapmgr.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chipaca) [reply overlord/snapstate/snapmgr.go]: sigh. of course `doSwitchSnap` exists already and is what's called from `Switch`, whereas `switch-snap-channel` is called from update to switch not only the tracked channel etc, but also the current channel.
I'll refactor and add comments so I don't forget.

File: daemon/api.go (hunk 1 above)
Comment (chipaca) [reply daemon/api.go]: no it wasn't (in my defense, before this PR there were no tests for switch in daemon)

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -1314,30 +1314,23 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
-
-	// switched the tracked channel
-	snapst.Channel = snapsup.Channel
-	// optionally support switching the current snap channel too, e.g.
-	// if a snap is in both stable and candidate with the same revision
-	// we can update it here and it will be displayed correctly in the UI
-	if snapsup.SideInfo.Channel != "" {
-		snapst.CurrentSideInfo().Channel = snapsup.Channel
-	}
+type switchFlags struct {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: doSwitchFlags

File: overlord/snapstate/handlers.go [hunk 10]
```diff
@@ -1314,30 +1314,23 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
+	return m.genericSwitchSnap(t, switchFlags{switchCurrentChannel: true})
 }
 
+// doSwitchSnap switches the snap's tracking channel and/or cohort, *without*
+// switching the current snap channel. For use from 'Switch'.
 func (m *SnapManager) doSwitchSnap(t *state.Task, _ *tomb.Tomb) error {
+	return m.genericSwitchSnap(t, switchFlags{})
+}
+
+func (m *SnapManager) genericSwitchSnap(t *state.Task, flags switchFlags) error {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: s/genericSwitchSnap/genericDoSwithSnap/

File: overlord/snapstate/snapstate.go [hunk 11]
```diff
@@ -1209,7 +1215,29 @@ func resolveChannel(st *state.State, snapName, newChannel string, deviceCtx Devi
...
-// Switch switches a snap to a new channel
+func switchSummary(snap, chanFrom, chanTo, cohFrom, cohTo string) string {
+	switchChannel := chanFrom != chanTo && chanTo != ""
+	switchCohort := cohFrom != cohTo && cohTo != ""
+	switch {
+	case switchChannel && !switchCohort:
+		return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q"), snap,
+			chanFrom, chanTo)
+	case switchCohort && !switchChannel:
+		return fmt.Sprintf(i18n.G("Switch snap %q from cohort %q to %q"), snap,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: if cohFrom is empty this is a bit strange.

"from no cohort to cohort %q" ?

File: overlord/snapstate/snapstate.go [hunk 12]
```diff
@@ -1209,7 +1215,29 @@ func resolveChannel(st *state.State, snapName, newChannel string, deviceCtx Devi
...
+			strutil.ElliptRight(cohFrom, 10),
+			strutil.ElliptRight(cohTo, 10),
+		)
+	default:
+		// either they both change or they both don't change (the latter might actually be an error)
+		return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and from cohort %q to %q"), snap,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: same

File: overlord/snapstate/snapstate.go [hunk 13]
```diff
@@ -1242,20 +1270,30 @@ func Switch(st *state.State, name string, opts *RevisionOptions) (*state.TaskSet
 
 	snapsup := &SnapSetup{
 		SideInfo:    snapst.CurrentSideInfo(),
-		Channel:     opts.Channel,
 		InstanceKey: snapst.InstanceKey,
+		CohortKey:   snapst.CohortKey,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: could we have here as well a comment about defaulting to original values?

File: daemon/api.go (hunk 1 above)
Comment (mvo5) [reply daemon/api.go]: *cough* no tests for switch at all - oh well - I think that was my fault :( Thanks for fixing this!

**Pull request merged**